lib/minify.js

```diff
@@ -0,0 +1,168 @@
+                toplevel: toplevel,
+                bare_returns: options.parse ? options.parse.bare_returns : undefined
+            });
+        }
+        if (!options.fromString) files = exports.simple_glob(files);
+        [].concat(files).forEach(function (files, i) {
+            if (typeof files === 'string') {
+                addFile(files, options.fromString ? i : files);
+            } else {
+                for (var fileUrl in files) {
+                    addFile(files[fileUrl], fileUrl);
+                }
+            }
+        });
+    }
+    if (options.wrap) {
+      toplevel = toplevel.wrap_commonjs(options.wrap, options.exportAll);
+    }
+
+    // 2. compress
+    if (options.compress) {
+        var compress = { warnings: options.warnings };
+        merge(compress, options.compress);
+        toplevel.figure_out_scope(options.mangle);
+        var sq = Compressor(compress);
+        toplevel = sq.compress(toplevel);
+    }
+
+    // 3. mangle properties
+    if (options.mangleProperties || options.nameCache) {
+        options.mangleProperties = options.mangleProperties || {};
+        options.mangleProperties.cache = readNameCache(options.nameCache, "props");
+        toplevel = mangle_properties(toplevel, options.mangleProperties);
+        writeNameCache(options.nameCache, "props", options.mangleProperties.cache);
+    }
+
+    // 4. mangle
+    if (options.mangle) {
+        toplevel.figure_out_scope(options.mangle);
+        toplevel.compute_char_frequency(options.mangle);
+        toplevel.mangle_names(options.mangle);
+    }
+
+    // 5. output
+    var inMap = options.inSourceMap;
+    var output = {};
+    if (typeof options.inSourceMap == "string") {
+        inMap = JSON.parse(exports.readFile(options.inSourceMap));
+    }
+    if (options.outSourceMap || options.sourceMapInline) {
+        output.source_map = SourceMap({
+            // prefer outFileName, otherwise use outSourceMap without .map suffix
+            file: options.outFileName || (typeof options.outSourceMap === 'string' ? options.outSourceMap.replace(/\.map$/i, '') : null),
+            orig: inMap,
+            root: options.sourceRoot
+        });
+        if (options.sourceMapIncludeSources) {
+            for (var file in sourcesContent) {
+                if (sourcesContent.hasOwnProperty(file)) {
+                    output.source_map.get().setSourceContent(file, sourcesContent[file]);
+                }
+            }
+        }
+
+    }
+    if (options.output) {
+        merge(output, options.output);
+    }
+    var stream = OutputStream(output);
+    toplevel.print(stream);
+
+    var source_map = output.source_map;
+    if (source_map) {
+        source_map = source_map + "";
+    }
+
+    var mappingUrlPrefix = "\n//# sourceMappingURL=";
+    if (options.sourceMapInline) {
+        stream += mappingUrlPrefix + "data:application/json;charset=utf-8;base64," + new Buffer(source_map).toString("base64");
```

Comment:
@avdg The use of `Buffer` is not portable in the browser. Can a `base64_decode` function be exported much like `exports.readFile` and `exports.writeFile` and the code changed to be something like the following?

```
     if (options.sourceMapInline && exports.base64_decode) {
         stream += mappingUrlPrefix + "data:application/json;charset=utf-8;base64," + exports.base64_decode(source_map);
```

See: https://github.com/mishoo/UglifyJS2/pull/1490#issuecomment-280067609

/cc @alexlamsl